File: mac_scripts/functions/git-tools/ssh-url.mjs
// Last path segment is the repo; preceding segments (groups/subgroups) are owner.
const SSH_RE = /^git@([^:]+):(.+)\/([^/]+?)(?:\.git)?$/i;

/**
 * @param {string} value
 * @returns {string}
 */
export function slugifySegment(value) {
  return value
    .toLowerCase()
    .replace(/[^a-z0-9-]+/g, "-")
    .replace(/-+/g, "-")
    .replace(/^-|-$/g, "");
}

/**
 * @param {string} owner
 * @param {string} repo
 * @returns {string}
 */
export function buildProjectName(owner, repo) {
  return `${slugifySegment(owner)}-${slugifySegment(repo)}`;
}

/**
 * @param {string} input
 * @returns {{ ok: true, host: string, owner: string, repo: string, projectName: string }
 *   | { ok: false, error: string }}
 */
export function parseSshGitUrl(input) {
  if (typeof input !== "string" || !input.trim()) {
    return { ok: false, error: "Invalid SSH URL" };
  }

  const trimmed = input.trim();

  if (/^https?:\/\//i.test(trimmed)) {
    return { ok: false, error: "HTTPS URLs are not supported" };
  }

  const match = trimmed.match(SSH_RE);
  if (!match) {
    return { ok: false, error: "Invalid SSH URL" };
  }

  const [, host, owner, repo] = match;

  return {
    ok: true,
    host,
    owner,
    repo,
    projectName: buildProjectName(owner, repo),
  };
}

/**
 * @param {string} input
 * @returns {{ ok: true, canonical: string, sshUrl: string }
 *   | { ok: false, error: string }}
 */
export function canonicalizeSshGitUrl(input) {
  const parsed = parseSshGitUrl(input);
  if (!parsed.ok) {
    return parsed;
  }

  const canonical = `git@${parsed.host.toLowerCase()}:${parsed.owner}/${parsed.repo}`;
  return {
    ok: true,
    canonical,
    sshUrl: `${canonical}.git`,
  };
}

File: mac_scripts/functions/git_tools/ssh_url.py
from dataclasses import dataclass
import re

_SSH_RE = re.compile(r"^git@([^:]+):(.+)/([^/]+?)(?:\.git)?$", re.IGNORECASE)

@dataclass(frozen=True, slots=True)
class ParsedSshUrl:
    host: str
    owner: str
    repo: str
    ssh_url: str
    canonical: str

    @property
    def project_name(self) -> str:
        return build_project_name(self.owner, self.repo)

def slugify_segment(value: str) -> str:
    return re.sub(r"-+", "-", re.sub(r"[^a-z0-9-]+", "-", value.lower())).strip("-")

def build_project_name(owner: str, repo: str) -> str:
    return f"{slugify_segment(owner)}-{slugify_segment(repo)}"

def parse_ssh_git_url(value: str) -> ParsedSshUrl:
    """Parse a Git SSH URL, preserving owner/repo casing and path nesting.

    The JavaScript implementation reports these two validation failures with
    stable messages; ValueError carries those messages at the Python boundary.
    """

    if not isinstance(value, str) or not value.strip():
        raise ValueError("Invalid SSH URL")

    trimmed = value.strip()
    if re.match(r"^https?://", trimmed, re.IGNORECASE):
        raise ValueError("HTTPS URLs are not supported")

    match = _SSH_RE.match(trimmed)
    if match is None:
        raise ValueError("Invalid SSH URL")

    host, owner, repo = match.groups()
    canonical = f"git@{host.lower()}:{owner}/{repo}"
    return ParsedSshUrl(
        host=host,
        owner=owner,
        repo=repo,
        ssh_url=f"{canonical}.git",
        canonical=canonical,
    )

def canonicalize_ssh_git_url(value: str) -> ParsedSshUrl:
    return parse_ssh_git_url(value)
